File: p31.py
def coins(n, total, denoms):
    if len(denoms) == 0:
        return total
    coin = denoms[0]
    rest = denoms[1:]
    multiplier = 0
    while n >= coin * multiplier:
        if n == coin * multiplier:
            total += 1
            break
        total = coins(n - coin * multiplier, total, rest)
        multiplier += 1
    return total

print coins(200, 0, [1,2,5,10,20,50,100,200])

File: p31.hs
module Main where

-- With at least mult of the first coin in the list, how many ways can
-- we make the target?

coins target denoms mult
    | denoms == []                 = 0
    | mult * currentCoin > target  = 0
    | mult * currentCoin == target = 1
    | otherwise                    =
        -- this multiplier with the rest of the coins
        (coins (target - (mult * currentCoin)) (tail denoms) 0) +
        -- increase the multiplier
        coins target denoms (mult + 1)
    where
      currentCoin = head denoms

main = do
  putStrLn $ show $ coins 200 [200,100,50,20,10,5,2,1] 0
